### DISCUSSION TITLE
Broadcast with provided buffers

### DISCUSSION THREAD

**Discussion by erenon** [axboe/liburing#1266] [Q&A] [answered]
Hi, I'd like to send (broadcast) the same buffer to many (20-100) sockets, as quickly as possible. Using a simple send operation, I can set the same buffer for each socket. However, to avoid accidental reordering of sends ([see this discussion](https://github.com/axboe/liburing/discussions/1102)), I'm serializing the sends user-space. To make this more efficient, hoping I can avoid userspace-kernelspace copies, I'm looking at provided buffers. Questions:

 1. Is it correct, that each outgoing socket needs its own provided buffer ring? (Otherwise I can't concurrently call send on them)
 2. Can I somehow add the same buffer to many buffer rings, or is there another way to send the same buffer to many sockets w/o reordering?

I already tried send_zerocopy with registered buffers, with no clear performance improvement. Thanks!

**Comment by axboe** [comment, marked as answer]
1. Yes if you went that route, you'd need a ring per socket for that, as each send would consume the provided buffer.
2. You can add the same buffer(s) to as many rings as you'd like, they don't have to be separate buffers.

For zero copy, recent kernels do much better here - the networking stack had some trouble properly coalescing for io_uring zerocopy send. 6.11 might look a whole lot better for that.

**Reply by erenon** [reply]
Thanks!
